homeassistant/components/recorder/__init__.py

```diff
@@ -305,8 +305,37 @@ def _setup_connection(self):
         models.Base.metadata.create_all(self.engine)
         session_factory = sessionmaker(bind=self.engine)
         Session = scoped_session(session_factory)
+        self._migrate_schema()
         self.db_ready.set()
 
+    def _migrate_schema(self):
```

Comment:
Before our SQL Alchemy migration, we kept a table with the last migrations that we applied. That way we can do a simple check.

Pseudo Python:

```python
last_migration = getModel('schema_migration').last

if not last_migration:
    # Do first migration
if last_migration < 2:
    # Do second migration
if last_migration < 3:
    # Do third migration

getModel('schema_migration').insert(3)
```

That way startup is relatively fast because we only need to check a single value instead of all indexes on a specific table.